Changes most gas reaction rates to be proportional to moles. Goodbye fixed gas reaction rates. Proto-nitrate BZ response now has a chance to shoot nuclear particles.

Pickle-Coding opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
This changes the reaction rate of most gas reactions. Most gas reactions will now have consumption rates proportional to how much there is. Standardises the reactions to have a minimal mole factor, which is the minimum of the consumed gases divided by their consumption rates, then multiply them by efficiency variables that range from 0 to 1, with 1 being 100% efficiency. This PR also reduces the range of proto-nitrate bz response radiation pulses, but now they have a chance to shoot nuclear particles.

### Changed Reactions
#### N2O Formation
Heat based reaction. Efficiency drops to 0% when it's at its lower or upper temperature limits, and reaches 100% efficiency at 225 Kelvin.
#### Pluoxium Formation (via CO2 + O2 + trit reaction. Supermatter + CO2 + O2  reaction is untouched)
Pressure based reaction. Converges to 100% efficiency as pressure rises.
#### Nitrium Formation
Heat based reaction. Converges to 100% efficiency as temperature rises.
#### Nitrium Decomposition
Heat based reaction. Higher temperature increases efficiency, but doesn't converge to 100%. More oxygen will increase the reaction rate.
#### Halon Combustion
Heat based reaction. Converges to 100% efficiency as temperature rises.
#### Healium Formation
Pressure based reaction. Converges to 100% efficiency as pressure rises.
#### Zauker Formation
Pressure based reaction. Converges to 100% efficiency as pressure rises.
#### Proto-Nitrate Formation
Heat based reaction. The efficiency reaches 100% when the temperature is either 7,500 Kelvin, but drops to 0% efficiency when the temperature is 5,000 Kelvin, or 10,000 Kelvin.
#### Proto-Nitrate Hydrogen Response
Pressure based reaction. Converges to 100% efficiency as pressure rises. The reaction efficiency is also dependent on how much proto-nitrate there is in relation to hydrogen.
#### Proto-Nitrate BZ Response
Reaction rate didn't change, but now there's a change to shoot a nuclear particle. The higher the reaction rate, the higher the chance is to shoot a nuclear particle. The range of the radiation pulse has been halved. Nuclear particles will have the same volume scaling requirement as radiation pulses do.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Quite a significant amount of gas reactions do not scale with moles or volume, or even worse, have a fixed and arbitrary reaction rate, and this creates a problem where it's almost always faster to spread the reaction to multiple gasmixtures, usually open turfs. Open turf reactions are cool, but not so cool when it's the fastest method for almost every gas reaction.

The current gas reaction rates are also kind of lame and boring. Giving reactions their own optimal or suboptimal conditions will make the process of gas synthesis a little more interesting than just needing to know the temperature range and gas consumption rates, then spreading it to as many turfs as you can. Even though it complicates the reactions, it will make atmospherics a little more intuitive for people that don't code dive, as making a lot of gases react in a small environment makes more sense than spreading the environment to separate gas mixtures to get past the arbitrary reaction rate.

Proto-nitrate BZ response and tritium response are a little too similar except for the gases they use and their output, so I'm going to give bz response a reaction some people may enjoy. It has the same volume scaling requirements as radiation pulses do, so don't expect some nerd to dump it in waste then irradiate everyone.
<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
balance: Most gas reaction rates have been changed. They are now based on moles and efficiency, where efficiency is a multiplier that ranges from 0 to 1.
balance: N2O Formation is now heat and molar based. Most efficient at 225 Kelvin.
balance: Pluoxium Formation is now pressure and molar based. More efficient at higher pressures.
balance: Nitrium Formation is now heat and molar based. More efficient at higher temperatures.
balance: Nitrium Decomposition is now heat (not ordinary efficiency scale) and molar based. More efficient at higher temperatures and oxygen.
balance: Halon Combustion is now heat and molar based. More efficient at higher temperatures.
balance: Healium Formation is now pressure and molar based. More efficient at higher pressures.
balance: Zauker Formation is now pressure and molar based. More efficient at higher pressures.
balance: Proto-Nitrate Formation is now temperature and molar based. Max efficiency at 7,500 Kelvin, but halts when temperature reaches formation barriers..
balance: Proto-Nitrate Hydrogen Response is now pressure based. More efficient at higher pressures. More proto-nitrate increases the efficiency.
balance: Halves the radiation pulse range of Proto-Nitrate BZ Response.
add: Proto-Nitrate BZ Response now has a chance to shoot nuclear particles the higher the reaction rate is!
fix: Fixed the Proto-Nitrate Tritium Response requiring the Proto-Nitrate BZ Response energy threshold to release radiation, and vise-versa.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Pickle-Coding committed:
Why?

For some reason the sign on the if check was reversed, which causes freon burn rate to be based on freon moles instead of oxygen moles if there is lacking oxygen for full burn.

Not sure why it got flipped, but I'm just flipping the freon burn rates to correct it, so it should be based on oxygen moles if there isn't enough oxygen for fullburn.

Pickle-Coding committed:
Fixes freon hotspots killing themselves.

For some weird reason, the reaction would return no reaction if the holder isn't a turf. This caused hotspots caused by freon fires to kill themselves because there was no reaction results after hotspot called react on a gasmixture that isn't in a turf.

This fixes that by removing the check at the start to check if the holder is a turf, and instead just check if it's a turf for hot ice spawning.

Pickle-Coding committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into freon_combustion_reaction_rate_thing

Pickle-Coding committed:
No more static gas reaction rates.

Changes the gas reaction efficiency of most gas reactions. The reaction rates now scales with moles. Most instances of static gas reaction rates have been removed.

Standardises the reactions to use the minimum moles of the involved gases for the reaction efficiency. The minimum moles will get multiplied by some efficiency variables that range from 0 to 1, depending on the gas reaction.

Heat based gas reactions will either have a heat efficiency that has an optimal point(s), or the heat efficiency will converge to 1 as temperature rises. Some heat based reactions have been replaced with pressure based reactions, usually higher pressure will mean an efficiency variable that converges to 1.

Gas reactions that uses a catalyst will have an efficiency variable that is based on how much of the catalyst there is in relation to the other gases of the reactions. Example is N2O formation. The efficiency rises the more BZ there is in relation to oxygen or nitrogen (minimal_mole_factor).

Proto-nitrate BZ response radiation pulse strength has been reduced by 75%, but now it has a chance to shoot a nuclear particle, based on reaction efficiency.

Pickle-Coding committed:
Reaction factors descriptions.

Adds descriptions for gas reactions in reaction_factors.dm to update with the new gas reaction efficiency changes.

Pickle-Coding committed:
Correct a number in reaction_factors.

SnoopCooper commented:
Finally, no more free reaction rates from valves or multiple tiles

zxaber commented:
>Pluoxium Formation
Pressure based reaction. Converges to 100% efficiency as pressure rises.


RIP CO2 engines having a neat byproduct.

Pickle-Coding commented:
This doesn't change the CO2 supermatter interaction. This just changes the auxiliary pluoxium formation reaction.

Pickle-Coding commented:
Unless you mean the reaction change will make the byproduct look less unique.

I think I'll tone down the pluoxium formation reaction rate a bit, so CO2 in supermatter will be more favourable.

zxaber commented:
I meant the former, as I asumed this meant the normally lower engine chamber pressure would kill the reaction.

I don't really mind there even being a "proper" way to make Pluox, it's just neat that you can use the supermatter as a catalyst for making other gases.

SnoopCooper commented:
I think we should keep the regular pluox reaction. It’s needed for proto nitrate and such so it probably shouldn’t be locked behind the SM. Also one of the more useful gasses you can make.

Pickle-Coding commented:
New pluox formation reaction can be faster than the current one if you pressurise it.

Pickle-Coding committed:
Tones down PN_BZASE nuclear particle chance.

The consumed_amount will need to be 20 times higher to achieve the same effect.

Pickle-Coding committed:
lol

For some reason the radiation release energy threshold for proto-nitrate bz and tritium response was swapped around. This commit will swap them back to where they're supposed to be.

Pickle-Coding commented:
I'll simplify the proto-nitrate formation reaction when I have the time.

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/reaction_factors.dm:
```diff
@@ -90,6 +90,7 @@
 		/datum/gas/tritium = "Tritium is consumed at 0.01 reaction rate",
 		/datum/gas/pluoxium = "Pluoxium is produced at 1 reaction rate",
 		/datum/gas/hydrogen = "Hydrogen is produced at 0.01 reaction rate",
+		"Pressure" = "The higher the pressure the faster the reaction rate goes.",
```
```suggestion
		"Pressure" = "Reaction rate scales directly with pressure",
```
Something like this?

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/reaction_factors.dm:
```diff
@@ -100,13 +101,13 @@
 	factor = list(
-		/datum/gas/oxygen = "[MINIMUM_MOLE_COUNT] moles of oxygen need to be present for the reaction to occur. Not consumed.",
+		/datum/gas/oxygen = "[MINIMUM_MOLE_COUNT] moles of oxygen need to be present for the reaction to occur. Not consumed. Increases reaction rate.",
 		/datum/gas/nitrium = "Nitrium is consumed at 1 reaction rate",
 		/datum/gas/hydrogen = "Hydrogen is produced at 1 reaction rate",
 		/datum/gas/nitrogen = "Nitrogen is produced at 1 reaction rate",
```
imo we prob shouldn't be mentioning reaction rate as a unit in our uis like this, but that can get handled later

Pickle-Coding replied:
I suppose it could be something like "X is produced per mol of Y consumed" or something like that. Probably out of scope for this PR though.

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/reaction_factors.dm:
```diff
@@ -148,7 +149,8 @@
 		/datum/gas/freon = "Freon is consumed at 2.75 reaction rate",
 		/datum/gas/healium = "Healium is produced at 3 reaction rate",
-		"Temperature" = "Can only occur between [HEALIUM_FORMATION_MIN_TEMP] - [HEALIUM_FORMATION_MAX_TEMP]. Higher temperature increases the reaction rate.",
+		"Pressure" = "The higher the pressure the faster the reaction rate goes.",
```
See before? maybe? this is a little stilted, but I honestly don't really mind that much if you wanna keep it, I've got a weird sense for this stuff anyway

Pickle-Coding replied:
I'll improve it later today when I have time.

Pickle-Coding replied:
Changed.

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/reaction_factors.dm:
```diff
@@ -175,33 +178,35 @@
 		/datum/gas/proto_nitrate = "Proto-Nitrate is produced at 2.2 reaction rate",
 		"Energy" = "[PN_FORMATION_ENERGY] joules of energy is released per reaction rate",
-		"Temperature" = "Can only occur between [PN_FORMATION_MIN_TEMPERATURE] - [PN_FORMATION_MAX_TEMPERATURE] kelvin. Higher temperature increases the reaction rate.",
+		"Temperature" = "Can only occur between [PN_FORMATION_MIN_TEMPERATURE] - [PN_FORMATION_MAX_TEMPERATURE] kelvin. Reaction rate peaks at [PN_FORMATION_LOWER_TEMPERATURE_MAXIMUM_EFFICIENCY] and [PN_FORMATION_UPPER_TEMPERATURE_MAXIMUM_EFFICIENCY] Kelvin, and drops at [PN_FORMATION_MIN_TEMPERATURE], 7,500, and [PN_FORMATION_MAX_TEMPERATURE] Kelvin.",
```
This is really runon, what's it trying to get across?

Pickle-Coding replied:
I'll simplify the reaction itself in the next commit, along with description.

Pickle-Coding replied:
The reaction itself and thhe description has been simplified.

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -409,9 +406,9 @@
 	var/list/cached_gases = air.gases
 	if(cached_gases[/datum/gas/oxygen][MOLES] < cached_gases[/datum/gas/freon][MOLES] * FREON_OXYGEN_FULLBURN)
-		freon_burn_rate = (cached_gases[/datum/gas/freon][MOLES] / FREON_BURN_RATE_DELTA) * temperature_scale
-	else
 		freon_burn_rate = ((cached_gases[/datum/gas/oxygen][MOLES] / FREON_OXYGEN_FULLBURN) / FREON_BURN_RATE_DELTA) * temperature_scale
+	else
+		freon_burn_rate = (cached_gases[/datum/gas/freon][MOLES] / FREON_BURN_RATE_DELTA) * temperature_scale
```
what's goin on here?

Pickle-Coding replied:
The branch of this PR was created from the freon fire fix PR, though I'm not sure why it is showing those changes when the PR already got merged.

LemonInTheDark replied:
you should pull tgstation master into your branch, see if that fixes it

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -458,16 +455,20 @@
 /datum/gas_reaction/nitrousformation/react(datum/gas_mixture/air)
 	var/list/cached_gases = air.gases
-	var/heat_efficency = min(cached_gases[/datum/gas/oxygen][MOLES], cached_gases[/datum/gas/nitrogen][MOLES] * INVERSE(2))
+	var/temperature = air.temperature
+	var/minimal_mole_factor = min(cached_gases[/datum/gas/oxygen][MOLES], cached_gases[/datum/gas/nitrogen][MOLES] * INVERSE(2))
+	var/thermal_efficiency = 1 - ((temperature - N2O_FORMATION_OPTIMAL_TEMPERATURE) / N2O_FORMATION_TEMPERATURE_SCALE) ** 2
+	var/bz_efficiency = 1 - NUM_E ** -(thermal_efficiency * cached_gases[/datum/gas/bz][MOLES] / minimal_mole_factor)
+	var/heat_efficency = min(minimal_mole_factor * bz_efficiency, cached_gases[/datum/gas/oxygen][MOLES], cached_gases[/datum/gas/nitrogen][MOLES] * INVERSE(2))
```
Jesus this is complex, what's going on here exactly?
Also why do you calculate minimal_mole_factor's value twice?

Pickle-Coding replied:
thermal_efficiency is just a parabola that peaks at 1 when between minimum and maximum temperature.

The equation for bz_efficiency and other variables that look similar could be understood as an integral of e^-x, where x is whatever I write in the parenthesis. This would evaluate as progressing roughly 63.212% towards 100% every time x increases by 1. Since the slope of the equation is 1 when x is 0, it can be more easily understood as efficiency being slightly below x when x is a low value (I'd say below 0.2), so the efficiency would be almost 10% if x is 0.1.

I'm not sure if my explanation could be understood easily, but hopefully understanding that the efficiency would be near x when x is a low value, then having noticeable diminishing returns when x increases to high values (above 0.2) wouldn't be too problematic to maintain. This formula is very similar (exactly in probably one case) to what people would use if they wanted to calculate the chance of an event occuring where the same chance roll occurs multiple times, though I don't know if that has a name.

In this case for bz_efficiency formula, the efficiency rises as how much bz there is in relation to the minimum moles of other gases mapped with consumption ratios. Thermal efficiency dictates how efficient the bz would be. If thermal efficiency is at 100%, and there is 1 part bz per 5 part minimal moles, then you can assume that heat_efficiency would be almost 1/5th of minimal_moles.

Now for why minimal_moles gets calculated twice, the first one is used as part to calculate the production/consumption/heat_efficiency rates. The second one is redundant as mathematically this shouldn't go above minimum moles of any gas, but is used anyways to err on the side of caution, as with every gas reaction.

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -608,7 +609,10 @@
 /datum/gas_reaction/pluox_formation/react(datum/gas_mixture/air, datum/holder)
 	var/list/cached_gases = air.gases
-	var/produced_amount = min(PLUOXIUM_FORMATION_MAX_RATE, cached_gases[/datum/gas/carbon_dioxide][MOLES], cached_gases[/datum/gas/oxygen][MOLES] * INVERSE(0.5), cached_gases[/datum/gas/tritium][MOLES] * INVERSE(0.01))
+	var/pressure = air.return_pressure()
+	var/minimal_mole_factor = min(cached_gases[/datum/gas/carbon_dioxide][MOLES], cached_gases[/datum/gas/oxygen][MOLES] * INVERSE(0.5), cached_gases[/datum/gas/tritium][MOLES] * INVERSE(0.01))
+	var/pressure_efficiency = 1 - NUM_E ** -(pressure / PLUOXIUM_FORMATION_PRESSURE_SCALE)
+	var/produced_amount = min(minimal_mole_factor * pressure_efficiency, cached_gases[/datum/gas/carbon_dioxide][MOLES], cached_gases[/datum/gas/oxygen][MOLES] * INVERSE(0.5), cached_gases[/datum/gas/tritium][MOLES] * INVERSE(0.01))
```
same questions here

Pickle-Coding replied:
Mostly same as above, but I'll give a short explanation for each of these.

Reaches 1 - inverse e efficiency when pressure reaches pressure scale. If it's 20% of pressure scale or below, you can safely assume it is near the pressure over pressure scale.

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -646,16 +650,18 @@
 	var/list/cached_gases = air.gases
 	var/temperature = air.temperature
-	var/heat_efficency = min(temperature / NITRIUM_FORMATION_TEMP_DIVISOR, cached_gases[/datum/gas/tritium][MOLES], cached_gases[/datum/gas/nitrogen][MOLES], cached_gases[/datum/gas/bz][MOLES] * INVERSE(0.05))
+	var/minimal_mole_factor = min(cached_gases[/datum/gas/tritium][MOLES], cached_gases[/datum/gas/nitrogen][MOLES], cached_gases[/datum/gas/bz][MOLES] * INVERSE(0.05))
+	var/thermal_efficiency = 1 - NUM_E ** -((temperature - NITRIUM_FORMATION_MIN_TEMP) / NITRIUM_FORMATION_TEMP_DIVISOR)
+	var/heat_efficency = min(minimal_mole_factor * thermal_efficiency, cached_gases[/datum/gas/tritium][MOLES], cached_gases[/datum/gas/nitrogen][MOLES], cached_gases[/datum/gas/bz][MOLES] * INVERSE(0.05))
```
and here

Pickle-Coding replied:
Same as above, but for temperature. It's offset by the minimum temperature for formation, so it is easier to understand this as the difference between temperature and minimum formation temperature when you want to scale it.

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -836,8 +842,10 @@
 	var/list/cached_gases = air.gases
 	var/temperature = air.temperature
+	var/minimal_mole_factor = min(cached_gases[/datum/gas/halon][MOLES], cached_gases[/datum/gas/oxygen][MOLES] * INVERSE(20))
+	var/thermal_efficiency = 1 - NUM_E ** -((temperature - FIRE_MINIMUM_TEMPERATURE_TO_EXIST) / HALON_COMBUSTION_TEMPERATURE_SCALE)
 
-	var/heat_efficency = min(temperature / ( FIRE_MINIMUM_TEMPERATURE_TO_EXIST * 10), cached_gases[/datum/gas/halon][MOLES], cached_gases[/datum/gas/oxygen][MOLES] * INVERSE(20))
+	var/heat_efficency = min(thermal_efficiency * minimal_mole_factor / 10, cached_gases[/datum/gas/halon][MOLES], cached_gases[/datum/gas/oxygen][MOLES] * INVERSE(20))
```
You know what just like

Pickle-Coding replied:
Same as above.

Pickle-Coding committed:
Update code/modules/atmospherics/gasmixtures/reaction_factors.dm

Co-authored-by: LemonInTheDark <58055496+LemonInTheDark@users.noreply.github.com>

Pickle-Coding committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into gamering

Pickle-Coding committed:
I am not sure if this will do anything.

Pickle-Coding committed:
This is the commit that actually does things.

Replaces a magic number I forgot to replace with a define.

Better reaction factor descriptions.

Proto-nitrate formation is more simple. Peaks at 7,500 Kelvin.

Changes peak style heat efficiency formula defines to be dynamic with other defines, so they don't need to be changed when other defines get changed.

Significantly simplifies catalyst based reactions, so that they just use how many moles there is of the catalyst instead of efficiency formula. This is because the efficiency would progress quickly as reaction continues, and would finish shortly after the difference between using the formula and just using moles would be noticeable.

Reverts zauker decomposition changes, as the change I mentioned above would just turn this to a static reaction rate that only scales with volume, and that would just be an irrelevant balance change outside the scope of the PR.

Maybe other stuff I forgot about.

Pickle-Coding committed:
Forgot to change description of these.

Pickle-Coding committed:
Bye.

Removes unused define.

No, this is not an empty commit, Git extension.

Pickle-Coding committed:
Exponential cumulative macro.

Adds EXPONENTIAL_CUMULATIVE macro, and replaces all instances of cumulative exponential distribution with the macro.

Pickle-Coding commented:
> This PR has been inactive for long enough to be automatically marked as stale. This means it is at risk of being auto closed in ~ 7 days, please address any outstanding review items and ensure your PR is finished, if these are all true and you are auto-staled anyway, you need to actively ask maintainers if your PR will be merged. Once you have done any of the previous actions then you should request a maintainer remove the stale label on your PR, to reset the stale timer. If you feel no maintainer will respond in that time, you may wish to close this PR youself, while you seek maintainer comment, as you will then be able to reopen the PR yourself

![image](https://user-images.githubusercontent.com/58013024/170800671-2af8f7a4-ee35-4413-a6a2-ad36b46a0be3.gif)

LemonInTheDark commented:
When you say scales to 100%, does that mean one tick of processing will fully complete the reaction

Pickle-Coding commented:
Scales 100% for the variable, which gets multiplied by minimal moles. This could fully complete some reactions instantly if someone somehow got the requirements, but some reactions have divisors that divide after the 100% scaling variables, so some reactions won't be able to be completed instantly.

LemonInTheDark commented:
Why is it good for reactions to be completed instantly? or well, close to it

Pickle-Coding commented:
Ideally I am aiming for a 5-20% reaction rate. I could pribably cap the reaction rates to loke 50% or something if you think people would be able to complete it instantly too easily.

Pickle-Coding committed:
It was annoying.

Changes EXPONENTIAL_CUMULATIVE macro to only use one parameter, who cares about the second parameter?

Changes most of the changed gas reactions to not consume more than half of their gas in any situation.

Pickle-Coding commented:
I've changed it so most of the reactions won't be able to consume more than half of their gas. Left N2O formation (which I assume is supposed to be a fast reaction as it instantly completes before this PR) and proto-nitrate hydrogen response out for now because of the way they scale.

LemonInTheDark commented:
being honest with you, I'm not really comfortable with making things this proportional. It's a sweeping change, and doing it all at once like this makes it really hard to judge what exactly is changing.
That may just be a me problem though, it's hard for me to tell. I'd be more ok if you just made them scale consistently, so I didn't need to worry about balancing "optimized" reactions and such. Also I don't really wanna make gas reader pages even harder to read.

BZ projectiles is just cool, feel free to repr
